tools/clang/lib/SPIRV/TypeTranslator.cpp

```diff
@@ -1139,13 +1148,31 @@ TypeTranslator::getLayoutDecorations(const DeclContext *decl, LayoutRule rule,
     std::tie(memberAlignment, memberSize) =
         getAlignmentAndSize(fieldType, rule, &stride);
 
+    // The next avaiable location after layouting the previos members
+    const uint32_t nextLoc = offset;
+
     alignUsingHLSLRelaxedLayout(fieldType, memberSize, &memberAlignment,
                                 &offset);
 
-    // Each structure-type member must have an Offset Decoration.
-    if (const auto *offsetAttr = field->getAttr<VKOffsetAttr>())
+    // The vk::offset attribute takes precedence over all.
+    if (const auto *offsetAttr = field->getAttr<VKOffsetAttr>()) {
       offset = offsetAttr->getOffset();
+    }
+    // The :packoffset() annotation takes precedence over normal layout
+    // calculation.
+    else if (const auto *pack = getPackOffset(declDecl)) {
+      const uint32_t packOffset =
+          pack->Subcomponent * 16 + pack->ComponentOffset * 4;
```

Comment:
should the multiplier always be 16, even if the struct contains no vectors?

Reply:
Yes, I think so. `c[subcomponent].[componet]` basically means place at the `[component]`th float in the `[subcomponent]`th float4. It's just a way of saying a location using float4 as basic unit; it has nothing to do with the underlying data type. A struct can also be placed at, say the 5th float4's location.

FXC allows to have `:packoffset()` on struct members.

See https://msdn.microsoft.com/en-us/library/windows/desktop/dd607358(v=vs.85).aspx

There are also some existing checks for `:packoffset()`:

https://github.com/Microsoft/DirectXShaderCompiler/blob/93af33644e0218bfe9a55e80a1a6d39a781a44b8/tools/clang/lib/Sema/SemaHLSL.cpp#L8679-L8700

I